src/grpc/mod.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use futures::channel::oneshot;
+use futures::executor::block_on;
+use futures::prelude::*;
+use grpcio::{Environment, ServerBuilder};
+use std::error::Error;
+use std::io::Read;
+use std::str;
+use std::sync::Arc;
+use std::{io, thread};
+use string_error::into_err;
+
+pub mod grpc_keyprovider;
+
+use grpc_keyprovider::keyprovider;
+use grpc_keyprovider::keyprovider_grpc::*;
+use grpc_keyprovider::keyprovider_structs::*;
+
+use super::kbc_manager;
+
+#[derive(Clone)]
+pub struct KeyProvider;
+
+impl KeyProviderService for KeyProvider {
+    fn un_wrap_key(
+        &mut self,
+        ctx: ::grpcio::RpcContext,
+        req: keyprovider::keyProviderKeyWrapProtocolInput,
+        sink: ::grpcio::UnarySink<keyprovider::keyProviderKeyWrapProtocolOutput>,
+    ) {
+        // Deserialize gRPC input to get kbs_annotation_packet.
+        let (annotation, dc) =
+            match get_annotation_and_dc_from_input(req.KeyProviderKeyWrapProtocolInput) {
+                Ok((a, d)) => (a, d),
+                Err(e) => {
+                    println!(
+                        "Some error happend when get annotation and dc from input: {}",
+                        e
+                    );
+                    return ();
+                }
+            };
+
+        // Parse the DC parameter in gRPC input and get the KBC name to be selected.
+        // For details on DC parameter,
+        // please refer to kbc_get_name function in kbc_manager.rs.
+        let kbc_name_string: String = match kbc_manager::kbc_get_name(dc.Parameters) {
+            Ok(name) => name,
+            Err(e) => {
+                println!("Some error happend when get kbc name from Dc: {}", e);
+                return ();
+            }
+        };
+
+        // Select the KBC module instance,
+        // and pass the annotation to the KBC module for content parsing and field decryption.
+        let mut kbc_manager = match kbc_manager::KBC_MANAGER.lock() {
```

Comment:
Do we really need a mutex to unwrap a key? Will the state of the KBC or KBC Manager change?

Reply:
Thank you for reminding of this. It is necessary to use mutex, because the supported KBC instances will be created and registered in the management list of KBC manager during AA startup, resulting in the change of KBC manager. To avoid the risk of deadlock, we changed Mutex<KbcManager> to Arc<Mutex<KbcManager>>.